In [1]:
import pandas as pd

In [ ]:
df_11 = pd.read_csv('/content/drive/MyDrive/버스/2021년_버스노선별_정류장별_시간대별_승하차_인원_정보(11월).csv', encoding='cp949')
df_11.columns

/usr/local/lib/python3.7/dist-packages/IPython/core/interactiveshell.py:2718: DtypeWarning: Columns (4) have mixed types.Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


Index(['사용년월', '노선번호', '노선명', '표준버스정류장ID', '버스정류장ARS번호', '역명', '00시승차총승객수',
       '00시하차총승객수', '1시승차총승객수', '1시하차총승객수', '2시승차총승객수', '2시하차총승객수', '3시승차총승객수',
       '3시하차총승객수', '4시승차총승객수', '4시하차총승객수', '5시승차총승객수', '5시하차총승객수', '6시승차총승객수',
       '6시하차총승객수', '7시승차총승객수', '7시하차총승객수', '8시승차총승객수', '8시하차총승객수', '9시승차총승객수',
       '9시하차총승객수', '10시승차총승객수', '10시하차총승객수', '11시승차총승객수', '11시하차총승객수',
       '12시승차총승객수', '12시하차총승객수', '13시승차총승객수', '13시하차총승객수', '14시승차총승객수',
       '14시하차총승객수', '15시승차총승객수', '15시하차총승객수', '16시승차총승객수', '16시하차총승객수',
       '17시승차총승객수', '17시하차총승객수', '18시승차총승객수', '18시하차총승객수', '19시승차총승객수',
       '19시하차총승객수', '20시승차총승객수', '20시하차총승객수', '21시승차총승객수', '21시하차총승객수',
       '22시승차총승객수', '22시하차총승객수', '23시승차총승객수', '23시하차총승객수', '등록일자'],
      dtype='object')

In [ ]:
# # df_11 = pd.read_csv('/content/drive/MyDrive/버스/2021년_버스노선별_정류장별_시간대별_승하차_인원_정보(11월).csv', encoding='cp949')
# # df_10 = pd.read_csv('/content/drive/MyDrive/버스/2021년_버스노선별_정류장별_시간대별_승하차_인원_정보(10월).csv', encoding='cp949')
# # df_09 = pd.read_csv('/content/drive/MyDrive/버스/2021년_버스노선별_정류장별_시간대별_승하차_인원_정보(09월).csv', encoding='cp949')

# month = ['11','10','09','08','07','06','05','04','03','02','01']

# import sys
# mod = sys.modules[__name__]

# for i in month:
#   df = pd.read_csv('/content/drive/MyDrive/버스/2021년_버스노선별_정류장별_시간대별_승하차_인원_정보(%s월).csv' %i, encoding='cp949')
#   df['승차총승객수'] = 0
#   for n in range(6,54,2):
#     df['승차총승객수'] = df['승차총승객수'] + df.iloc[:,n]
#   df = df[['노선번호', '노선명', '표준버스정류장ID', '승차총승객수']]
#   df.drop_duplicates(subset = ['노선번호', '표준버스정류장ID'], inplace=True)
#   df = df[['노선번호', '표준버스정류장ID', '승차총승객수']]
#   df = df.set_index(['노선번호'])
#   df.columns = ['표준버스정류장ID', '2021%s' %i]
#   df.sort_values(by=['표준버스정류장ID', '노선번호'], ascending=True, inplace=True)
#   setattr(mod, 'df_{}'.format(i), df)

# df_12 = pd.read_csv('/content/drive/MyDrive/버스/2020년_버스노선별_정류장별_시간대별_승하차_인원_정보(12월).csv', encoding='cp949')
# df = df_12
# df['승차총승객수'] = 0
# for i in range(6,54,2):
#   df['승차총승객수'] = df['승차총승객수'] + df.iloc[:,i]

# df_12 = df[['노선번호', '노선명', '표준버스정류장ID', '승차총승객수']]
# df_12.drop_duplicates(subset = ['노선번호', '표준버스정류장ID'], inplace=True)
# df_12 = df_12[['노선번호', '표준버스정류장ID', '승차총승객수']]
# df_12 = df_12.set_index(['노선번호'])
# df_12.columns = ['표준버스정류장ID', '202012']
# df_12.sort_values(by=['표준버스정류장ID', '노선번호'], ascending=True, inplace=True)

In [ ]:
# df_11 = pd.read_csv('/content/drive/MyDrive/버스/2021년_버스노선별_정류장별_시간대별_승하차_인원_정보(11월).csv', encoding='cp949')
# df_10 = pd.read_csv('/content/drive/MyDrive/버스/2021년_버스노선별_정류장별_시간대별_승하차_인원_정보(10월).csv', encoding='cp949')
# df_09 = pd.read_csv('/content/drive/MyDrive/버스/2021년_버스노선별_정류장별_시간대별_승하차_인원_정보(09월).csv', encoding='cp949')

month = ['11','10','09','08','07','06','05','04','03','02','01']

import sys
mod = sys.modules[__name__]

for i in month:
  df = pd.read_csv('/content/drive/MyDrive/버스/2021년_버스노선별_정류장별_시간대별_승하차_인원_정보(%s월).csv' %i, encoding='cp949')
  df['승차총승객수'] = 0
  for n in range(6,54,2):
    df['승차총승객수'] = df['승차총승객수'] + df.iloc[:,n]
  df = df[['노선번호', '노선명', '표준버스정류장ID', '역명', '승차총승객수']]
  df.drop_duplicates(subset = ['노선번호', '표준버스정류장ID'], inplace=True)
  df = df[['노선번호', '표준버스정류장ID', '역명', '승차총승객수']]
  df = df.set_index(['노선번호'])
  df.columns = ['표준버스정류장ID', '역명', '2021%s' %i]
  df.sort_values(by=['표준버스정류장ID', '역명', '노선번호'], ascending=True, inplace=True)
  setattr(mod, 'df_{}'.format(i), df)

df_12 = pd.read_csv('/content/drive/MyDrive/버스/2020년_버스노선별_정류장별_시간대별_승하차_인원_정보(12월).csv', encoding='cp949')
df = df_12
df['승차총승객수'] = 0
for i in range(6,54,2):
  df['승차총승객수'] = df['승차총승객수'] + df.iloc[:,i]

df_12 = df[['노선번호', '노선명', '표준버스정류장ID', '역명', '승차총승객수']]
df_12.drop_duplicates(subset = ['노선번호', '표준버스정류장ID'], inplace=True)
df_12 = df_12[['노선번호', '표준버스정류장ID', '역명', '승차총승객수']]
df_12 = df_12.set_index(['노선번호'])
df_12.columns = ['표준버스정류장ID', '역명', '202012']
df_12.sort_values(by=['표준버스정류장ID', '역명', '노선번호'], ascending=True, inplace=True)

/usr/local/lib/python3.7/dist-packages/IPython/core/interactiveshell.py:2718: DtypeWarning: Columns (4) have mixed types.Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)
/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:30: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy


merge_outer 써서 데이터프레임 합치기

In [ ]:
m = pd.merge(df_12,df_01, how='outer',on=['노선번호', '표준버스정류장ID', '역명'])
m = pd.merge(m,df_02, how='outer',on=['노선번호', '표준버스정류장ID', '역명'])
m = pd.merge(m,df_03, how='outer',on=['노선번호', '표준버스정류장ID', '역명'])
m = pd.merge(m,df_04, how='outer',on=['노선번호', '표준버스정류장ID', '역명'])
m = pd.merge(m,df_05, how='outer',on=['노선번호', '표준버스정류장ID', '역명'])
m = pd.merge(m,df_06, how='outer',on=['노선번호', '표준버스정류장ID', '역명'])
m = pd.merge(m,df_07, how='outer',on=['노선번호', '표준버스정류장ID', '역명'])
m = pd.merge(m,df_08, how='outer',on=['노선번호', '표준버스정류장ID', '역명'])
m = pd.merge(m,df_09, how='outer',on=['노선번호', '표준버스정류장ID', '역명'])
m = pd.merge(m,df_10, how='outer',on=['노선번호', '표준버스정류장ID', '역명'])
m = pd.merge(m,df_11, how='outer',on=['노선번호', '표준버스정류장ID', '역명'])
m

표준버스정류장ID                  역명  202012  ...  202109  202110  202111
노선번호                                         ...                        
470   100000001             종로2가사거리  5036.0  ...  5687.0  6578.0  7328.0
741   100000001             종로2가사거리  4235.0  ...  5400.0  5587.0  6232.0
N37   100000001             종로2가사거리    46.0  ...    59.0    71.0   306.0
100   100000002         창경궁.서울대학교병원  2018.0  ...  2205.0  2423.0  3114.0
102   100000002         창경궁.서울대학교병원   741.0  ...   916.0   956.0  1150.0
...         ...                 ...     ...  ...     ...     ...     ...
5627  213000481    광명역데시앙.일직동행정복지센터     NaN  ...     NaN     NaN   660.0
5633  213000481    광명역데시앙.일직동행정복지센터     NaN  ...     NaN     NaN   590.0
3323  227000488  미사강변더샵센트럴포레.동원로얄듀크     NaN  ...     NaN     NaN   762.0
3323  227000632  미사강변더샵센트럴포레.동원로얄듀크     NaN  ...     NaN     NaN    85.0
2416  998501931       중랑공영차고지(기점가상)     NaN  ...     NaN     NaN     3.0

[40268 rows x 14 columns]

- 특수문자 인식 에러로 ?로 뜨는 결과 10개 확인하여 대체
- (가상) 이 들어가는 의미 홈페이지에 기재 x -> 승차인원 존재하기에 가상 다 지워버림

In [ ]:
m['역명'] = m['역명'].str.replace('?', '·')

In [ ]:
search = ['(가상)', '(종점가상)', '(기점가상)']
for i, word in enumerate(m['역명']):
  for j in search:
    if j in word: 
        m.iloc[i,1] = word.strip(j)
m

표준버스정류장ID                  역명  202012  ...  202109  202110  202111
노선번호                                         ...                        
470   100000001             종로2가사거리  5036.0  ...  5687.0  6578.0  7328.0
741   100000001             종로2가사거리  4235.0  ...  5400.0  5587.0  6232.0
N37   100000001             종로2가사거리    46.0  ...    59.0    71.0   306.0
100   100000002         창경궁.서울대학교병원  2018.0  ...  2205.0  2423.0  3114.0
102   100000002         창경궁.서울대학교병원   741.0  ...   916.0   956.0  1150.0
...         ...                 ...     ...  ...     ...     ...     ...
5627  213000481    광명역데시앙.일직동행정복지센터     NaN  ...     NaN     NaN   660.0
5633  213000481    광명역데시앙.일직동행정복지센터     NaN  ...     NaN     NaN   590.0
3323  227000488  미사강변더샵센트럴포레.동원로얄듀크     NaN  ...     NaN     NaN   762.0
3323  227000632  미사강변더샵센트럴포레.동원로얄듀크     NaN  ...     NaN     NaN    85.0
2416  998501931             중랑공영차고지     NaN  ...     NaN     NaN     3.0

[40268 rows x 14 columns]

null 값 처리
1. 모든 컬럼에서 null 값인거 제거해보았으나 없었음 (40268)
2. 가장 최근 데이터인 202111에서 null 값이 존재하면 안되므로 202111에 존재하는 null값 다 제거 (40268 -> 37176)
3. thresh=12 써서 12개월 중 1개월에서만 null값 보이는 것 빼고 null 값 다 제거 (37176 -> 34902)
4. null 값은 전후 달의 데이터 평균으로 채우기

In [ ]:
#1
m_drop = m.dropna(axis=0, how='all')
m_drop

표준버스정류장ID                  역명  202012  ...  202109  202110  202111
노선번호                                         ...                        
470   100000001             종로2가사거리  5036.0  ...  5687.0  6578.0  7328.0
741   100000001             종로2가사거리  4235.0  ...  5400.0  5587.0  6232.0
N37   100000001             종로2가사거리    46.0  ...    59.0    71.0   306.0
100   100000002         창경궁.서울대학교병원  2018.0  ...  2205.0  2423.0  3114.0
102   100000002         창경궁.서울대학교병원   741.0  ...   916.0   956.0  1150.0
...         ...                 ...     ...  ...     ...     ...     ...
5627  213000481    광명역데시앙.일직동행정복지센터     NaN  ...     NaN     NaN   660.0
5633  213000481    광명역데시앙.일직동행정복지센터     NaN  ...     NaN     NaN   590.0
3323  227000488  미사강변더샵센트럴포레.동원로얄듀크     NaN  ...     NaN     NaN   762.0
3323  227000632  미사강변더샵센트럴포레.동원로얄듀크     NaN  ...     NaN     NaN    85.0
2416  998501931             중랑공영차고지     NaN  ...     NaN     NaN     3.0

[40268 rows x 14 columns]

In [ ]:
#2
m_drop = m_drop.dropna(axis=0, how='any', subset=['202111'])
m_drop

표준버스정류장ID                  역명  202012  ...  202109  202110  202111
노선번호                                         ...                        
470   100000001             종로2가사거리  5036.0  ...  5687.0  6578.0  7328.0
741   100000001             종로2가사거리  4235.0  ...  5400.0  5587.0  6232.0
N37   100000001             종로2가사거리    46.0  ...    59.0    71.0   306.0
100   100000002         창경궁.서울대학교병원  2018.0  ...  2205.0  2423.0  3114.0
102   100000002         창경궁.서울대학교병원   741.0  ...   916.0   956.0  1150.0
...         ...                 ...     ...  ...     ...     ...     ...
5627  213000481    광명역데시앙.일직동행정복지센터     NaN  ...     NaN     NaN   660.0
5633  213000481    광명역데시앙.일직동행정복지센터     NaN  ...     NaN     NaN   590.0
3323  227000488  미사강변더샵센트럴포레.동원로얄듀크     NaN  ...     NaN     NaN   762.0
3323  227000632  미사강변더샵센트럴포레.동원로얄듀크     NaN  ...     NaN     NaN    85.0
2416  998501931             중랑공영차고지     NaN  ...     NaN     NaN     3.0

[37176 rows x 14 columns]

In [ ]:
#3
m_drop2 = m_drop.dropna(axis=0, thresh=13)
m_drop2

표준버스정류장ID             역명  202012  202101  ...  202108  202109  202110  202111
노선번호                                            ...                                
470   100000001        종로2가사거리  5036.0  5085.0  ...  6053.0  5687.0  6578.0  7328.0
741   100000001        종로2가사거리  4235.0  4086.0  ...  5529.0  5400.0  5587.0  6232.0
N37   100000001        종로2가사거리    46.0     NaN  ...    58.0    59.0    71.0   306.0
100   100000002    창경궁.서울대학교병원  2018.0  1939.0  ...  2131.0  2205.0  2423.0  3114.0
102   100000002    창경궁.서울대학교병원   741.0   815.0  ...   837.0   916.0   956.0  1150.0
...         ...            ...     ...     ...  ...     ...     ...     ...     ...
3412  124000216  강일리버파크3단지.1단지     NaN    10.0  ...   118.0    12.0    21.0    14.0
201   221000057        대림한숲아파트     NaN  1299.0  ...  1783.0  1808.0  1917.0  2009.0
201   221000102     LG원앙.한성아파트     NaN  1847.0  ...  2462.0  2448.0  2424.0  2435.0
201   221000110       대림.하나아파트     NaN   125.0  ...   250.0   288.0   241.0   234.0
0017  998601292        청암자이아파트     NaN     1.0  ...     1.0     4.0     3.0     0.0

[34902 rows x 14 columns]

In [ ]:
m_drop2.to_csv('/content/drive/Shareddrives/2021 팜 경진대회/3 코드/3 버스노선 관련/노선번호, 버스정류장별 승차인구 202012~202111.csv', encoding='cp949')



---


In [3]:
df = pd.read_csv('/content/drive/Shareddrives/2021 팜 경진대회/3 코드/3 버스노선 관련/노선번호, 버스정류장별 승차인구 202012~202111.csv', encoding='cp949')
df

노선번호  표준버스정류장ID             역명  202012  ...  202108  202109  202110  202111
0       470  100000001        종로2가사거리  5036.0  ...  6053.0  5687.0  6578.0  7328.0
1       741  100000001        종로2가사거리  4235.0  ...  5529.0  5400.0  5587.0  6232.0
2       N37  100000001        종로2가사거리    46.0  ...    58.0    59.0    71.0   306.0
3       100  100000002    창경궁.서울대학교병원  2018.0  ...  2131.0  2205.0  2423.0  3114.0
4       102  100000002    창경궁.서울대학교병원   741.0  ...   837.0   916.0   956.0  1150.0
...     ...        ...            ...     ...  ...     ...     ...     ...     ...
34897  3412  124000216  강일리버파크3단지.1단지     NaN  ...   118.0    12.0    21.0    14.0
34898   201  221000057        대림한숲아파트     NaN  ...  1783.0  1808.0  1917.0  2009.0
34899   201  221000102     LG원앙.한성아파트     NaN  ...  2462.0  2448.0  2424.0  2435.0
34900   201  221000110       대림.하나아파트     NaN  ...   250.0   288.0   241.0   234.0
34901  0017  998601292        청암자이아파트     NaN  ...     1.0     4.0     3.0     0.0

[34902 rows x 15 columns]

In [4]:
df.isnull().sum()

노선번호           0
표준버스정류장ID      0
역명             0
202012       277
202101       684
202102       208
202103         0
202104         1
202105         0
202106         0
202107         2
202108         1
202109         4
202110         0
202111         0
dtype: int64

In [5]:
for i in range(len(df)):
  row = df.iloc[i]
  row_isna = pd.DataFrame(row.isna())
  for j in range(len(row_isna)):
    if row_isna.iloc[j].bool():
      try:
        df.iloc[i, j] = mean(df.iloc[i, j-1], df.iloc[i, j+1])
      except:
        df.iloc[i, j] = df.iloc[i, j+1]

In [6]:
df.isnull().sum()

노선번호         0
표준버스정류장ID    0
역명           0
202012       0
202101       0
202102       0
202103       0
202104       0
202105       0
202106       0
202107       0
202108       0
202109       0
202110       0
202111       0
dtype: int64

In [8]:
# 202012 의 경우 좌측이 없으므로 우측 데이터와 동일하게,
# 그외는 좌우 평균값으로 결측치를 대체
df.to_csv('/content/drive/Shareddrives/2021 팜 경진대회/3 코드/3 버스노선 관련/노선별 버스정류장별 승차인구 202012~202111_최종.csv', encoding='cp949')